File: problems/0003_longest-substring-without-repeating-characters/solution1.rb
#!/usr/bin/ruby

# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    max, start, e = [0, 0, 0]
    for i in 0..(s.length - 1)
        idx = find(s, s[i], start, e)
        e += 1
        if idx == nil
            max = [max, e - start].max
        else
            start = idx + 1
        end
    end
    return max
end

def find(s, c, start, e)
    for i in start..(e - 1)
        if s[i] == c
            return i
        end
    end
    return nil
end

File: problems/0003_longest-substring-without-repeating-characters/Solution1.py
#!/usr/bin/python3

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        m, start, end = 0, 0, 0
        for c in s:
            idx = s.find(c, start, end)
            end += 1
            if idx == -1:
                m = max(m, end - start)
            else:
                start = idx + 1
        return m
